File: Labs/experiment1/G_simple_calculator.py
import sys

class Calculator:
    def __init__(self):
        self.opstk = []
        self.numstk = []
        self.pri = {'+': 1, '-': 1, '*': 2, '/': 2}
        self.ans1 = -1
        self.ans2 = -1
        self.opsz = 0
        self.numsz = 0
        self.flag = True
        self.ans = ""

    def cal(self, op):
        num2 = self.numstk.pop()
        num1 = self.numstk.pop()
        if op == '+':
            self.numstk.append(num1 + num2)
        elif op == '-':
            self.numstk.append(num1 - num2)
        elif op == '*':
            self.numstk.append(num1 * num2)
        elif op == '/':
            if num2 == 0:
                print(f"error:{num1}/{num2}")
                self.flag = False
                return
            self.numstk.append(num1 // num2)
        self.numsz -= 1

    def solve(self):
        data = sys.stdin.read()
        i = 0
        while i < len(data):
            c = data[i]
            if c.isspace():
                i += 1
                continue
            if c.isdigit():
                self.ans += c
            else:
                if self.ans:
                    self.numstk.append(int(self.ans))
                    self.numsz += 1
                    self.ans2 = max(self.ans2, self.numsz)
                    self.ans = ""
                if c == '(':
                    self.opstk.append(c)
                    self.opsz += 1
                    self.ans1 = max(self.ans1, self.opsz)
                elif c == ')':
                    while self.opstk[-1] != '(':
                        op = self.opstk.pop()
                        self.opsz -= 1
                        self.cal(op)
                        if not self.flag:
                            return
                    self.opstk.pop()
                    self.opsz -= 1
                else:
                    while (self.opstk and self.opstk[-1] != '('
                           and self.pri[self.opstk[-1]] >= self.pri[c]):
                        op = self.opstk.pop()
                        self.opsz -= 1
                        self.cal(op)
                        if not self.flag:
                            return
                    self.opstk.append(c)
                    self.opsz += 1
                    self.ans1 = max(self.ans1, self.opsz)
            i += 1
        if self.ans:
            self.numstk.append(int(self.ans))
            self.numsz += 1
            self.ans2 = max(self.ans2, self.numsz)
        while self.opstk:
            op = self.opstk.pop()
            self.opsz -= 1
            self.cal(op)
            if not self.flag:
                return

calc = Calculator()
calc.solve()
if calc.flag:
    print(calc.numstk[-1])
    print(calc.ans1, calc.ans2)

File: Labs/experiment1/G_simple_calculator.cpp
#include <iostream>
#include <stack>
#include <map>
using namespace std;

stack<char> opstk;
stack<long long> numstk;
map<char, int> pri;
int ans1 = -1, ans2 = -1, opsz = 0, numsz = 0;
bool last = 1, flag = 1;
string ans = "";

long long getnum(const string& s) {
    long long a = 0;
    for (char ch : s) a = a * 10 + ch - '0';
    return a;
}

void cal(char op) {
    long long num2 = numstk.top(); numstk.pop();
    long long num1 = numstk.top(); numstk.pop();
    if (op == '+') numstk.push(num1 + num2);
    else if (op == '-') numstk.push(num1 - num2);
    else if (op == '*') numstk.push(num1 * num2);
    else if (op == '/') {
        if (num2 == 0) {
            cout << "error:" << num1 << "/" << num2 << endl;
            flag = 0;
            return;
        }
        numstk.push(num1 / num2);
    }
    numsz--;
}

void solve() {
    char c;
    while (cin >> c) {
        if (c >= '0' && c <= '9') {
            ans += c;
            last = 0;
        } else {
            last = 1;
            if (ans != "") {
                numstk.push(getnum(ans));
                numsz++;
                ans2 = max(ans2, numsz);
                ans = "";
            }
            if (c == '(') {
                opstk.push(c);
                opsz++;
                ans1 = max(ans1, opsz);
            } else if (c == ')') {
                while (opstk.top() != '(') {
                    char op = opstk.top(); opstk.pop(); opsz--;
                    cal(op);
                    if (!flag) return;
                }
                opstk.pop(); opsz--;
            } else {
                while (!opstk.empty() && opstk.top() != '(' && pri[opstk.top()] >= pri[c]) {
                    char op = opstk.top(); opstk.pop(); opsz--;
                    cal(op);
                    if (!flag) return;
                }
                opstk.push(c);
                opsz++;
                ans1 = max(ans1, opsz);
            }
        }
    }
    if (ans != "") {
        numstk.push(getnum(ans));
        numsz++;
        ans2 = max(ans2, numsz);
    }
    while (!opstk.empty()) {
        char op = opstk.top(); opstk.pop(); opsz--;
        cal(op);
        if (!flag) return;
    }
}

int main() {
    pri['+'] = 1; pri['-'] = 1;
    pri['*'] = 2; pri['/'] = 2;
    solve();
    if (flag) {
        cout << numstk.top() << endl;
        cout << ans1 << " " << ans2 << endl;
    }
    return 0;
}
